middleware/request/retry.go

```diff
@@ -8,21 +8,27 @@ import (
 )
 
 // NewRetryMiddleware returns a new retry middleware that wraps the provided `next` request handler.
-// The middleware retries the request a maximum of `maxRetries` times with a delay of `interval` between each retry.
+// The middleware accepts 2 params consisting of:
+//   - `getRetryInterval`: A higher order function that returns a function to calculate next retry interval
+//   - `maxRetries`: maximum number of retries allowed for this middleware
+//
 // If the request succeeds at any retry, the middleware returns `nil`. If all retries fail, it returns the last error encountered.
-func NewRetryMiddleware(maxRetries int, interval time.Duration) func(next wasabi.RequestHandler) wasabi.RequestHandler {
+func NewRetryMiddleware(getRetryInterval GetRetryInterval, maxRetries int) func(next wasabi.RequestHandler) wasabi.RequestHandler {
 	return func(next wasabi.RequestHandler) wasabi.RequestHandler {
 		return dispatch.RequestHandlerFunc(func(conn wasabi.Connection, req wasabi.Request) error {
 			var err error
-			ticker := time.NewTicker(interval)
+
+			ticker := time.NewTicker(getRetryInterval(0))
+
```

Comment:
@shan-96  for exponential backoff this line will panic https://cs.opensource.google/go/go/+/refs/tags/go1.22.4:src/time/tick.go;l=21-23

Reply:
@ksysoev `0` is not the interval, its the iteration. If it'd have panicked, then this [test](https://github.com/shan-96/wasabi/blob/95-retry-backoff/middleware/request/retry_policy_test.go#L20) would have failed

Reply:
Please correct me if I'm wrong but This test is not testing execution of middleware. It's only testing calculations of interval, no time.Ticker is created over there. Based on the code 0 in any power will be 0

Reply:
Read the test one more time, you're probably right... it looks it should return non zero interval for 0 attempt, probably I miss understood the code. Let me double check it tomorrow and merge the pr. sorry for confusion

Reply:
Yeah , my bad I thought attempt is first argument of power function, but it's second